### PR TITLE
Add onnx export with verification

### PR DESCRIPTION
**Pull request by liqunfu**
Fixes # Issue#TBD.

### Description
A MONAI Bundle can contain models in PyTorch, TorchScript, and ONNX formats.
Utility exists in MONAI to export a model in TorchScript format. Exported TorchScript models can be verified with Torch runtime. 
It is helpful (and complete) to provide a same utility to export ONNX models and to validate the models with PyTorch, ONNX(https://github.com/onnx/onnx), and onnxruntime (https://github.com/microsoft/onnxruntime).

With the proposed utility, ONNX model can be added to bundles in model-zoo with calling monai.bundle ckpt_export.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

`./runtests.sh --quick --unittests  --disttests`
TOTAL                                                               72191  11290    84%

### CODE REVIEW

File: monai/networks/utils.py
Comment (ericspod) [inline monai/networks/utils.py:558]: With `convert_to_torchscript` we have the option to save to a file, can we add that here?

File: tests/test_convert_to_onnx.py
Comment (ericspod) [inline tests/test_convert_to_onnx.py:30]: `UNet` should be compatible with ONNX but not all models will be so we'll have to have per-model tests like we do with Torchscript. This can be left to future PRs if it's not a priority however.

**Comment by binliunls** [comment]
Hi @liqunfu ,
     Do you have an ETA for this PR? We are working on this too. If you can have this PR merged this week, we can wait. Otherwise we can start a PR for this and invite you to review it.

Thanks,
Bin

**Comment by liqunfu** [comment]
> Hi @liqunfu , Do you have an ETA for this PR? We are working on this too. If you can have this PR merged this week, we can wait. Otherwise we can start a PR for this and invite you to review it.
> 
> Thanks, Bin

I am starting today to update my PR with reviewers' comments and to fix CI failures. I was trying to add more test for models that are used by model-zoo but it is time consuming because I need to get evaluation data some of which also need to wait for access approval. I agree with @ericspod that it can be in another PR.

BTW thanks all for reviewing this PR. I will update it by the end of today.

**Comment by liqunfu** [comment]
> > Hi @liqunfu , Do you have an ETA for this PR? We are working on this too. If you can have this PR merged this week, we can wait. Otherwise we can start a PR for this and invite you to review it.
> > Thanks, Bin
> 
> I am starting today to update my PR with reviewers' comments and to fix CI failures. I was trying to add more test for models that are used by model-zoo but it is time consuming because I need to get evaluation data some of which also need to wait for access approval. I agree with @ericspod that it can be in another PR.
> 
> BTW thanks all for reviewing this PR. I will update it by the end of today.

@binliunls Please feel free to add to this PR with what you may already have so we can merge sooner.

File: .github/workflows/pythonapp.yml
Comment (Nic-Ma) [inline .github/workflows/pythonapp.yml:153]: @wyli I think we don't need to install here?

Thanks.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:993]: tag should be "onnx_export"

**Review by Nic-Ma — COMMENTED** [review]
Hi @liqunfu ,

About adding a new optional dependency package, please refer to:
https://github.com/Project-MONAI/MONAI/blob/dev/CONTRIBUTING.md#adding-new-optional-dependencies

Thanks.

**Comment by Nic-Ma** [comment]
Mark: This PR will be targeting MONAI 1.2 P1 feature as several bundles which can't be supported by `torch-trt` so far can leverage `ONNX path` to convert to TensorRT model.

Thanks.

File: monai/networks/utils.py
Comment (liqunfu) [reply monai/networks/utils.py:557]: updated the function so that it can be called with a filename. Converted onnx model will be saved to the file if provided.

File: tests/test_convert_to_onnx.py
Comment (liqunfu) [reply tests/test_convert_to_onnx.py:43]: I added test case for SegResNet. More test cases shall be added in subsequent PRs.

File: .github/workflows/pythonapp.yml
Comment (liqunfu) [reply .github/workflows/pythonapp.yml]: you are correct @Nic-Ma , I have the line removed - got to know it more.

File: monai/bundle/scripts.py
Comment (liqunfu) [reply monai/bundle/scripts.py]: Thanks for suggesting the lazy import. I have updated to use it. Also updated to onnx-export.

**Comment by liqunfu** [comment]
> Mark: This PR will be targeting MONAI 1.2 P1 feature as several bundles which can't be supported by `torch-trt` so far can leverage `ONNX path` to convert to TensorRT model.
> 
> Thanks.

I am very glad that onnx and onnxruntime can offer its value for MONAI.

File: monai/bundle/scripts.py
Comment (binliunls) [reply monai/bundle/scripts.py]: Thanks for updating. I think the model saving part in the `_export` function blocks you from using it. Please go ahead with your way. I can make a PR to update it later.

Thanks in advance.

File: monai/networks/utils.py
Comment (binliunls) [inline monai/networks/utils.py:607]: Hi @liqunfu ,
     Could you please also add the `dynamic_axes` parameter to the `convert_to_onnnx` function and the `torch.onnx_export`, in case the different batch sizes are needed like in the `trt_export` API.

Thanks,
Bin

File: tests/test_convert_to_onnx.py
Comment (ericspod) [reply tests/test_convert_to_onnx.py:43]: We can test the networks like we do for Torchscript, that is testing with random data on small versions of the networks since we're not concerned with results. Testing the network as they are in core suffices for testing the bundles that use them, you don't need to worry about testing bundles directly.

Comment (liqunfu) [reply tests/test_convert_to_onnx.py:43]: Thanks @ericspod , I created test_onnx_save test utility and applied it to TestRetinaNet. It does no cross validation between PyTorch and onnxruntime.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:1017]: Hi @binliunls ,

Please make below logic as a private function:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/scripts.py#L1132-L1135
And enable all the export APIs to get input shape from metadata if `input_shape` is None and `use_trace=True`.
And update the doc-string to make the logic clear.

Thanks.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:995]: Why you ignored the `key_in_ckpt` and `use_trace` args here?

Thanks.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:1003]: the default should be `False`?

Thanks.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:1025]: `use_trace=use_trace_` is enough as we already set default in the `pop` call above.

Thanks.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:955]: I don't think ONNX export has the logic "metadata and config included as JSON files" as described.

Thanks.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:974]: provide default values for `meta_file`, `config_file`net_id` and override pairs` -> "provide default values for all the parameters of this function".
Please also update all the export functions for this arg.

Thanks.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:980]: Please import ONNX in the `convert_to_onnx` instead.

Thanks.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:951]: Please add `converter_kwargs` parameter to pass all the other args for the `convert_to_onnx()` except for `model, inputs, use_trace`.
Please also add it to `ckpt_export()` and `trt_export()`.

Thanks.

**Review by Nic-Ma — COMMENTED** [review]
Hi @wyli ,

Do we need to add
```
onnx
onnxruntime; python_version <= '3.10'
```
to all the required places by: https://github.com/Project-MONAI/MONAI/blob/dev/CONTRIBUTING.md#adding-new-optional-dependencies?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @binliunls ,

Please help address the comments.

Thanks in advance.

File: docs/requirements.txt
Comment (liqunfu) [reply docs/requirements.txt:39]: onnxruntime does not have release package for python 3.11. https://pypi.org/project/onnxruntime/#files

File: monai/bundle/scripts.py
Comment (liqunfu) [reply monai/bundle/scripts.py:1003]: PyTorch trace is in general more reliable than scripting. scripting is useful for loop and if which is no comment in computer vision. But if it is a convention in MONAI, we can change default to False.

File: monai/bundle/scripts.py
Comment (liqunfu) [reply monai/bundle/scripts.py:1017]: > Hi @binliunls ,
> 
> Please help address the comments.
> 
> Thanks in advance.

Thank you @Nic-Ma  and @binliunls !

File: monai/bundle/scripts.py
Comment (Nic-Ma) [reply monai/bundle/scripts.py:1003]: I would suggest to use `jit.script()` if possible, especially for dynamic input shapes.
@ericspod Do you have any comments?

Thanks.

File: monai/bundle/scripts.py
Comment (binliunls) [reply monai/bundle/scripts.py:980]: Hi @Nic-Ma ,
     The last line in this API using the onnx to save models.  I think it is necessary here, unless we have a saver to replace this one.

Thanks,
Bin

Comment (Nic-Ma) [reply monai/bundle/scripts.py:980]: OK, how about moving it to the top of this file?
I think the reason why we import `torch-trt` in the function is just a workaround for the import issue that changes CUDA GPU index. @wyli .

Thanks.

File: tests/test_convert_to_onnx.py
Comment (binliunls) [reply tests/test_convert_to_onnx.py:43]: Hi @liqunfu ,
     Why do you add a `test_onnx_save` and apply it to TestRetinaNet. I think the `test_convert_to_onnx` and `test_bundle_onnx_export` is enough for this PR. We may add the bundle onnx test in MONAI model-zoo to test more models. Could you please remove these tests, if there are no other reasons?

Thanks,
Bin

Comment (liqunfu) [reply tests/test_convert_to_onnx.py:43]: As @ericspod commented, we need a way to test MONAI's nets with onnx in the same way nets are tested for TorchScript. For this purpose, RetinaNet is tested in test_retinanet via test_onnx (similar to test_script) by calling test_onnx_save. Eventually, all test_xxxnet.py (or similar) will have a test_onnx call to validate the net's definition with onnx export. It is for validation of individual torch nnmodules in MONAI.

On the other hand, test_convert_to_onnx is to test convert_to_onnx utility in case it is broken.

test_bundle_onnx_export (thank to you for adding it!) is to ensure the utility added to script (onnx_export) is working as we expected. so that we can use it to generate model bundles.

File: monai/bundle/scripts.py
Comment (ericspod) [reply monai/bundle/scripts.py:1030]: We should use `jit.script` where we can but there's times that we would want to use `jit.trace`. Not all networks are compatible with `jit.script` (or at least weren't in the past with previous versions of Pytorch) so we need that option, I'm not aware of any other reason to use one function or the other but they must exist. Networks created with `jit.script` appear compatible with ONNX from my testing.

**Comment by binliunls** [comment]
Hi @wyli ,
    The windows-latest ci/cd test reports OOM error. Could you please give me some advice on why this happened and how to fix it?

error:
```
======================================================================
ERROR: test_vnet_shape_5 (tests.test_vnet.TestVNet)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\parameterized\parameterized.py", line 620, in standalone_func
    return func(*(a + p.args), **p.kwargs, **kw)
  File "D:\a\MONAI\MONAI\tests\test_vnet.py", line 69, in test_vnet_shape
    net = VNet(**input_param).to(device)
  File "D:\a\MONAI\MONAI\monai\networks\nets\vnet.py", line 241, in __init__
    self.up_tr256 = UpTransition(spatial_dims, 256, 256, 2, act, dropout_prob=dropout_prob)
  File "D:\a\MONAI\MONAI\monai\networks\nets\vnet.py", line 151, in __init__
    self.ops = _make_nconv(spatial_dims, out_channels, nconvs, act)
  File "D:\a\MONAI\MONAI\monai\networks\nets\vnet.py", line 55, in _make_nconv
    layers.append(LUConv(spatial_dims, nchan, act, bias))
  File "D:\a\MONAI\MONAI\monai\networks\nets\vnet.py", line 36, in __init__
    self.conv_block = Convolution(
  File "D:\a\MONAI\MONAI\monai\networks\blocks\convolutions.py", line 143, in __init__
    conv = conv_type(
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\torch\nn\modules\conv.py", line 591, in __init__
    super(Conv3d, self).__init__(
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\torch\nn\modules\conv.py", line 137, in __init__
    self.weight = Parameter(torch.empty(
RuntimeError: [enforce fail at C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\c10\core\impl\alloc_cpu.cpp:72] data. DefaultCPUAllocator: not enough memory: you tried to allocate 32768000 bytes.

======================================================================
FAIL: test_cachedataset (tests.test_sampler_dist.DistributedSamplerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\MONAI\MONAI\tests\utils.py", line 539, in _wrapper
    assert results.get(), "Distributed call failed."
AssertionError: Distributed call failed.
```

Thanks,
Bin

File: monai/bundle/scripts.py
Comment (Nic-Ma) [reply monai/bundle/scripts.py]: Yes, that's why I would suggest to set the default to `False`, and I agree we need this parameter.

Thanks.

File: setup.cfg
Comment (Nic-Ma) [inline setup.cfg:81]: Hi @binliunls ,

Please add below things in this postion:
https://github.com/Project-MONAI/MONAI/blob/dev/setup.cfg#L135
```
nni =
    nni
optuna =
    optuna
onnx =
    onnx>=1.13.0
    onnxruntime
```

**Review by Nic-Ma — APPROVED** [review]
Looks good to me, put some minor comments inline.
@wyli Do you think we need to add all the optional packages in this function or only selected?
https://github.com/Project-MONAI/MONAI/blob/dev/monai/config/deviceconfig.py#L64

Thanks.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:58]: No need this `has_onnx` var?

Thanks.

File: docs/requirements.txt
Comment (binliunls) [reply docs/requirements.txt:39]: Hi @Nic-Ma @wyli ,
     What should I do with this `python_version`?

Thanks,
Bin

File: setup.cfg
Comment (Nic-Ma) [inline setup.cfg:81]: Hi @wyli , should we add " python_version <= '3.10'" here?

Thanks.

File: monai/bundle/scripts.py
Comment (Nic-Ma) [inline monai/bundle/scripts.py:1036]: Missed a blank: "to an onnx model.Should be a list like"

File: tests/test_bundle_onnx_export.py
Comment (Nic-Ma) [inline tests/test_bundle_onnx_export.py:28]: Hi @liqunfu ,

May I know why we skip tests on windows here?
I also see some CI failure about the RetinaNet tests on windows, could you please help take a look?

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Thanks for the update, looks good to me.
Please fix the failed CI tests on windows, if it's expected error, please add `skip_if_windows` decorator.

Thanks.

File: tests/test_bundle_onnx_export.py
Comment (liqunfu) [reply tests/test_bundle_onnx_export.py:28]: I see test_bundle_trt_export and test_bundle_ckpt_export are also skipped on Windows. I think the common reason they are skipped is because the way configuration files are parsed by subprocess.run. It fails with this argument, f"['{config_file}','{def_args_file}']"].
I did not fix this because it is a pre-existing issue, I followed the way with which trt and ckpt bundle tests were treated to annotate the test with @skip_if_windows. 

Let me check the CI failure.

Comment (binliunls) [reply tests/test_bundle_onnx_export.py:28]: Hi @liqunfu ,
    I have fixed the CI bug by add `atol=0.1` to the unit test. Currently the ci failure is more like a github issue.
    @wyli  Could you please take a look at the window-latest test? It blocked this PR from merging.

Thanks in advance.

Comment (binliunls) [reply tests/test_bundle_onnx_export.py:28]: Hi @liqunfu ,
     You are the contributor to onnx right? Do you have some knowledge about whether it's possible that the import of onnx and onnxruntime lead to the memory overhead issue on windows?

Thanks,
Bin

Comment (liqunfu) [reply tests/test_bundle_onnx_export.py:28]: Hi @binliunls @wyli @Nic-Ma , 

It turned out that the CI failure was caused by a bug in pythonapp.yml (https://github.com/Project-MONAI/MONAI/pull/6237/commits/3a80a9810aeaca84bbe4af1c7e86061af2eb4f96). I made a correction and the CI pipeline is passing.
I see all required pipelines have passed. Hope the rest will complete without any issue.
Please let me know if there is anything I need to do to have this PR merged.

Thank you,
Liqun

Comment (wyli) [reply tests/test_bundle_onnx_export.py:28]: thanks for looking into this @liqunfu, I didn't pay attention to the page file config issue https://github.com/al-cheb/configure-pagefile-action/issues/5 it's been there for a while, interesting that we didn't notice this issue earlier...I'll merge this for now and verify it with all the postmerge tests.

Comment (Nic-Ma) [reply tests/test_bundle_onnx_export.py:28]: @liqunfu Thanks very much for your help here!

Comment (binliunls) [reply tests/test_bundle_onnx_export.py:28]: @liqunfu , thank you for your your help!

**Pull request merged**
